### DISCUSSION TITLE
CI tests start to fail after MinGW was updated from 11.2.0 to 12.2.0

### DISCUSSION THREAD

**Discussion by TheMostDiligent** [actions/virtual-environments#8372] [General questions] [answered]
Hello!

After the most recent runner image update, our test applications [fail to start](https://github.com/DiligentGraphics/DiligentCore/actions/runs/6291828159/job/17083207437#step:6:56) on Windows-2022 image the following error code:

```
Process completed with exit code 127.
```

The only thing that changed appears to be the MinGW version, which was updated from 11.2.0 to 12.2.0.
It worked fine in version  20230910.1.0 and fails in version 20230918.1.0.

I am very puzzled by this issue as I am not even sure what the actual problem is. The file is 100% there and I can dump all the DLLs needed by it:

```
ntdll.dll => /c/Windows/SYSTEM[32](https://github.com/DiligentGraphics/DiligentCore/actions/runs/6291828159/job/17083207437#step:6:34)/ntdll.dll (0x7ff9f0dc0000)
	KERNEL32.DLL => /c/Windows/System32/KERNEL32.DLL (0x7ff9f0450000)
	KERNELBASE.dll => /c/Windows/System32/KERNELBASE.dll (0x7ff9ee5b0000)
	apphelp.dll => /c/Windows/SYSTEM32/apphelp.dll (0x7ff9e9b20000)
	comdlg32.dll => /c/Windows/System32/comdlg32.dll (0x7ff9efee0000)
	msvcrt.dll => /c/Windows/System32/msvcrt.dll (0x7ff9efe30000)
	combase.dll => /c/Windows/System32/combase.dll (0x7ff9f0050000)
	ucrtbase.dll => /c/Windows/System32/ucrtbase.dll (0x7ff9eeb20000)
	RPCRT4.dll => /c/Windows/System32/RPCRT4.dll (0x7ff9f0720000)
	shcore.dll => /c/Windows/System32/shcore.dll (0x7ff9efaf0000)
	msvcp_win.dll => /c/Windows/System32/msvcp_win.dll (0x7ff9ee370000)
	USER32.dll => /c/Windows/System32/USER32.dll (0x7ff9f0bd0000)
	win32u.dll => /c/Windows/System32/win32u.dll (0x7ff9ee920000)
	GDI32.dll => /c/Windows/System32/GDI32.dll (0x7ff9f05a0000)
	gdi32full.dll => /c/Windows/System32/gdi32full.dll (0x7ff9ee490000)
	SHLWAPI.dll => /c/Windows/System32/SHLWAPI.dll (0x7ff9efc80000)
	SHELL32.dll => /c/Windows/System32/SHELL32.dll (0x7ff9eee20000)
	COMCTL32.dll => /c/Windows/WinSxS/amd64_microsoft.windows.common-controls_6595b64144ccf1df_5.82.20[34](https://github.com/DiligentGraphics/DiligentCore/actions/runs/6291828159/job/17083207437#step:6:36)8.1_none_22e4d0cd70d0f01d/COMCTL32.dll (0x7ff9c3df0000)
	ole32.dll => /c/Windows/System32/ole32.dll (0x7ff9f05e0000)
	ADVAPI32.dll => /c/Windows/System32/ADVAPI32.dll (0x7ff9efa00000)
	sechost.dll => /c/Windows/System32/sechost.dll (0x7ff9efbe0000)
	libwinpthread-1.dll => /mingw64/bin/libwinpthread-1.dll (0x7ff9ea300000)
	libgcc_s_seh-1.dll => /mingw64/bin/libgcc_s_seh-1.dll (0x7ff9e25[40](https://github.com/DiligentGraphics/DiligentCore/actions/runs/6291828159/job/17083207437#step:6:42)000)
	libstdc++-6.dll => /mingw64/bin/libstdc++-6.dll (0x7ff9d7240000)
```

When I build the executable locally with the same MinGW version, everything works just fine.

Any help would be highly appreciated!

**Comment by vpolikarpov-akvelon** [comment]
Hey @TheMostDiligent. I tried to investigate your issue and can confirm that it's MinGW update to blame because its rollback fixes workflow. But I failed to find a reason for that.

I tried to run you workflow with two different versions of MinGW:

- [11.2.0-12.0.1-9.0.0-r1](https://github.com/brechtsanders/winlibs_mingw/releases/tag/11.2.0-12.0.1-9.0.0-r1) was the version installed on runner prior to recent update - your workflow works with this version
- [12.2.0-14.0.6-10.0.0-ucrt-r2](https://github.com/brechtsanders/winlibs_mingw/releases/tag/12.2.0-14.0.6-10.0.0-ucrt-r2) is a version in between the old one and the new one  - your workflow fails with the same error as you observe

I don't understand what may be the reason. I would suggest you to read notes for both releases carefully. The only thing that I noticed is UCRT usage in the second one, may this affect your workflow?

**Reply by TheMostDiligent** [reply]
Thank you for looking into this!

The problem is that I can't reproduce the issue locally with the same version. I don't even understand why exactly the app fails to start.
My theory was that it can't find the DLLs for some reason, so I created an experimental workflow that dumps all dependencies, and everything [looks good](https://github.com/DiligentGraphics/DiligentCore/actions/runs/6291828159/job/17083207437#step:6:31).
Can you clarify which variation of the 12.2.0 mingw build you are using?

**Reply by vpolikarpov-akvelon** [reply]
As for MinGW 12.2.0, we are using this build currently: https://github.com/niXman/mingw-builds-binaries/releases/12.2.0-rt_v10-rev2/

**Reply by TheMostDiligent** [reply]
Right, but there are 4 variations (and even more variations in later releases):
* [x86_64-12.2.0-release-posix-seh-msvcrt-rt_v10-rev2.7z](https://github.com/niXman/mingw-builds-binaries/releases/download/12.2.0-rt_v10-rev2/x86_64-12.2.0-release-posix-seh-msvcrt-rt_v10-rev2.7z)
* [x86_64-12.2.0-release-posix-seh-ucrt-rt_v10-rev2.7z](https://github.com/niXman/mingw-builds-binaries/releases/download/12.2.0-rt_v10-rev2/x86_64-12.2.0-release-posix-seh-ucrt-rt_v10-rev2.7z)
* [x86_64-12.2.0-release-win32-seh-msvcrt-rt_v10-rev2.7z](https://github.com/niXman/mingw-builds-binaries/releases/download/12.2.0-rt_v10-rev2/x86_64-12.2.0-release-win32-seh-msvcrt-rt_v10-rev2.7z)
* [x86_64-12.2.0-release-win32-seh-ucrt-rt_v10-rev2.7z](https://github.com/niXman/mingw-builds-binaries/releases/download/12.2.0-rt_v10-rev2/x86_64-12.2.0-release-win32-seh-ucrt-rt_v10-rev2.7z)

Another question: release 13.1.0 has been available for a while now, so why are you not using the latest version?

**Reply by vpolikarpov-akvelon** [reply]
Oh, you are right. We install `x86_64-12.2.0-release-posix-seh-ucrt-rt_v10-rev2.7z`.

And I don't have answer about 13.1.0 right now, sorry. I'll let you know if I learn anything about it. But anyway I doubt that further upgrade is able to fix the issue.

**Reply by TheMostDiligent** [reply]
> We install x86_64-12.2.0-release-posix-seh-ucrt-rt_v10-rev2.7z.

When I try to build and run locally, everything works fine. The most plausible theory I have is that some required DLLs are not in the path:

```
Dump of file MinGW-12.2.0-posix-seh-ucrt\Tests\DiligentCoreTest\DiligentCoreTest.exe

File Type: EXECUTABLE IMAGE

  Image has the following dependencies:

    comdlg32.dll
    libgcc_s_seh-1.dll
    KERNEL32.dll
    api-ms-win-crt-convert-l1-1-0.dll
    api-ms-win-crt-environment-l1-1-0.dll
    api-ms-win-crt-filesystem-l1-1-0.dll
    api-ms-win-crt-heap-l1-1-0.dll
    api-ms-win-crt-locale-l1-1-0.dll
    api-ms-win-crt-math-l1-1-0.dll
    api-ms-win-crt-private-l1-1-0.dll
    api-ms-win-crt-runtime-l1-1-0.dll
    api-ms-win-crt-stdio-l1-1-0.dll
    api-ms-win-crt-string-l1-1-0.dll
    api-ms-win-crt-time-l1-1-0.dll
    ole32.dll
    libwinpthread-1.dll
    SHELL32.dll
    SHLWAPI.dll
    USER32.dll
    libstdc++-6.dll
```

These `api-ms-win-crt-*` dlls is what differs between 12.2.0 and 11.2.0. In 11.2.0, there was just a single `msvcrt.dll`

**Reply by TheMostDiligent** [reply]
I think that with the `msvcrt` version it will most likely work. Do you think it makes sense to use this version to be consistent with old behavior and be backwards compatible?

**Reply by quyykk** [reply]
> The most plausible theory I have is that some required DLLs are not in the path

I have a couple of actions that don't have this problem, they start just fine. I suspect it's not any DLLs that are missing, but some other package in earlier PATH that have their own copy of the msvcrt MinGW DLLs, and those get picked up when you try to execute your executable instead of the ones from MinGW directly (and the uscrt vs. msvcrt mismatch causes the failure).

Can you try executing `where.exe libstdc++-6.dll`?

**Reply by TheMostDiligent** [reply]
@quyykk I think your are right. I already use `ldd` to dump the dependent DLLS (see above). And it outputs:

```
	libwinpthread-1.dll => /mingw64/bin/libwinpthread-1.dll (0x7ff9ea300000)
	libgcc_s_seh-1.dll => /mingw64/bin/libgcc_s_seh-1.dll (0x7ff9e25[40](https://github.com/DiligentGraphics/DiligentCore/actions/runs/6291828159/job/17083207437#step:6:42)000)
	libstdc++-6.dll => /mingw64/bin/libstdc++-6.dll (0x7ff9d7240000)
```

However, the gcc compiler used is from a different path (it did not catch my attention):

```
C:/ProgramData/chocolatey/bin/gcc.exe
```

where.exe libstdc++-6.dll:

```
C:\Program Files\Git\mingw64\bin\libstdc++-6.dll
C:\Strawberry\c\bin\libstdc++-6.dll
C:\Strawberry\perl\bin\libstdc++-6.dll
```

So it is git whose mingw dlls are apparently used.

**Reply by TheMostDiligent** [reply]
@vpolikarpov-akvelon Where is mingw installed? Normally, the dlls are located next to the gcc. However, there are no dlls in `C:/ProgramData/chocolatey/bin`, where `gcc` is found:

```
/c/ProgramData/chocolatey/bin:
7z.exe
ISCC.exe
RefreshEnv.cmd
_processed.txt
addr2line.exe
ant.exe
ar.exe
aria2c.exe
as.exe
azcopy.exe
bazel.exe
bicep.exe
c++.exe
c++filt.exe
cc1.exe
cc1plus.exe
ccache-swig.exe
choco.exe
collect2.exe
cpp.exe
dlltool.exe
dllwrap.exe
docker-compose.exe
dwp.exe
elfedit.exe
f951.exe
fixincl.exe
g++-mapper-server.exe
g++.exe
gcc-ar.exe
gcc-nm.exe
gcc-ranlib.exe
gcc.exe
gcov-dump.exe
gcov-tool.exe
gcov.exe
gdb.exe
gdbm_dump.exe
gdbm_load.exe
gdbmtool.exe
gdborig.exe
gdbserver.exe
gendef.exe
genidl.exe
genpeimg.exe
gfortran.exe
gprof.exe
gradle.exe
helm.exe
hub.exe
jq.exe
julia.exe
kubectl-convert.exe
kubectl.exe
ld.bfd.exe
ld.exe
ld.gold.exe
lto-dump.exe
lto-wrapper.exe
lto1.exe
make.exe
mingw[32](https://github.com/DiligentGraphics/DiligentCore/actions/runs/6291828159/job/17200586349#step:6:34)-make.exe
minikube.exe
nm.exe
nssm.exe
nuget.exe
objcopy.exe
objdump.exe
packer.exe
python.exe
python3.9.exe
python3.exe
python3w.exe
pythonw.exe
ranlib.exe
readelf.exe
size.exe
strings.exe
strip.exe
swig.exe
vswhere.exe
widl.exe
windmc.exe
windres.exe
x86_[64](https://github.com/DiligentGraphics/DiligentCore/actions/runs/6291828159/job/17200586349#step:6:66)-w64-mingw32-c++.exe
x[86](https://github.com/DiligentGraphics/DiligentCore/actions/runs/6291828159/job/17200586349#step:6:88)_64-w64-mingw32-captoinfo.exe
x86_64-w64-mingw32-clear.exe
x86_64-w64-mingw32-g++.exe
x86_64-w64-mingw32-gcc-12.2.0.exe
x86_64-w64-mingw32-gcc-ar.exe
x86_64-w64-mingw32-gcc-nm.exe
x86_64-w64-mingw32-gcc-ranlib.exe
x86_64-w64-mingw32-gcc.exe
x86_64-w64-mingw32-gfortran.exe
x86_64-w64-mingw32-infocmp.exe
x86_64-w64-mingw32-infotocap.exe
x86_64-w64-mingw32-reset.exe
x86_64-w64-mingw32-tabs.exe
x86_64-w64-mingw32-tic.exe
x86_64-w64-mingw32-toe.exe
x86_64-w64-mingw32-tput.exe
x86_64-w64-mingw32-tset.exe
```
It looks like `libstdc++-6.dll` from mingw 12.2.0 can not be found.

**Reply by quyykk** [reply, marked as answer]
@TheMostDiligent Try `C:\ProgramData\chocolatey\lib\mingw\tools\install\mingw64\bin`, that's where they were last year,, maybe they're still there :smile:

**Reply by TheMostDiligent** [reply]
@quyykk You were right, the correct DLLs are apparently in this folder. Tweaking path as follows:
`PATH="/c/ProgramData/chocolatey/lib/mingw/tools/install/mingw64/bin:$PATH"`
makes the application start correctly.
Clearly, there is a misconfiguration, however this is at least a workaround.

**Comment by ChrisThrasher** [comment]
https://github.com/SFML/SFML/issues/2700

This problem has broken SFML's CI pipeline due to the MinGW 12 using UCRT instead of MSVCRT Libc. The only fix we can prove works is to manually downgrade back to the previous version of MinGW which still used MSVCRT libs. 

> The only thing that I noticed is UCRT usage in the second one, may this affect your workflow?

I'm quite certain this is the root of the problem.

**Comment by TheMostDiligent** [comment]
Since this turned out to be legitimate problem, I filed an [issue](https://github.com/actions/runner-images/issues/8399)
